core/io.go

```diff
@@ -0,0 +1,98 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"github.com/dicedb/dice/config"
+)
+
+type RESPParser struct {
+	c   io.ReadWriter
+	buf *bytes.Buffer
+}
+
+func NewRESPParser(c io.ReadWriter) *RESPParser {
+	return NewRESPParserWithBytes(c, []byte{})
+}
+
+func NewRESPParserWithBytes(c io.ReadWriter, initBytes []byte) *RESPParser {
+	var b []byte
+	var buf *bytes.Buffer = bytes.NewBuffer(b)
+	buf.Write(initBytes)
+	return &RESPParser{
+		c:   c,
+		buf: buf,
+	}
+}
+
+func (rp *RESPParser) DecodeOne() (interface{}, error) {
+	// assigning temporary buffer to read 512 bytes in one shot
+	// and reading them in a loop until we have all the data
+	// we want.
+	// note: the size 512 is arbitrarily chosen, and we can put
+	// a decent thought into deciding the optimal value (in case it affects the perf)
+	var buf []byte = make([]byte, config.IOBufferLength)
+
+	// keep reading the bytes from the buffer until the first /r is found
+	// note: there may be extra bytes read over the socket post /r
+	// but that is fine.
+	for {
+		n, err := rp.c.Read(buf)
+		if n <= 0 {
+			break
+		}
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		rp.buf.Write(buf[:n])
+
+		if bytes.Contains(buf, []byte{'\r', '\n'}) {
+			break
+		}
+	}
+
+	b, err := rp.buf.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+
+	switch b {
+	case '+':
+		return readSimpleString(rp.c, rp.buf)
+	case '-':
+		return readError(rp.c, rp.buf)
+	case ':':
+		return readInt64(rp.c, rp.buf)
+	case '$':
+		return readBulkString(rp.c, rp.buf)
+	case '*':
+		return readArray(rp.c, rp.buf, rp)
+	}
+
+	// this also captures the Cross Protocol Scripting attack.
+	// Since we do not support simple strings, anything that does
+	// not start with any of the above special chars will be a potential
+	// attack.
+	// Details: https://bou.ke/blog/hacking-developers/
+	return nil, errors.New("possible cross protocol scripting attack detected")
+}
+
+func (rp *RESPParser) DecodeMultiple() ([]interface{}, error) {
+	var values []interface{} = make([]interface{}, 0)
+	for {
+		value, err := rp.DecodeOne()
+		if err != nil {
+			return nil, err
+		}
+		values = append(values, value)
+		if rp.buf.Len() == 0 {
```

Comment:
Thanks for pointing it out. Here's my take on this

Command Pipelining is not a Transaction and the worst case here would be to perform another IO to read the remaining bytes.

Command Pipelining does not terminate with SPECIAL BYTE or EOF and hence there is no straight forward way to detect the end of it. If we make another `read` call on the socket we would remain blocked on the IO, again undesirable.

I think it is better to leave it as is because the worst case is just another IO to read subsequent commands.

Reply:
I feel I wasn't able to convey my point properly, I've made some changes to the go playground code to resemble to the code that we have: https://go.dev/play/p/mYYs3HnkURm

Please try executing this code, hope it'll clarify the point I'm trying to make.

**Expected Output**:
```
+PING
<nil>
+PING
<nil>
```

**Actual Output**:
```
+PING
<nil>
```


If we remove the block
```go
if bytes.Contains(tmp, []byte{'\r', '\n'}) {
	break
}
```

we get the expected output.

Reply:
Interesting discussion, @asutosh97. Though I would take the discussion in two ways - 

**1. Pipelining by following a valid RESP Spec-**

If you want to do two PING commands, a valid byte string as per RESP is coded here - https://go.dev/play/p/x6M802iFJRS

This is why you get proper output, verify here
```bash
(printf '*1\r\n$4\r\nPING\r\n*1\r\n$4\r\nPING\r\n';) | nc localhost 7379
+PONG
+PONG
```
The byte string you made is "+PING\r\n+PING\r\n". _This byte string actually breaks the RESP handler_ @arpitbbhayani in ``issue-51`` branch and in the current master branch as well. Since I find the fuzzing with protocols interesting and since **Cross Protocol Scripting** was my PR so I would like to strengthen RESP handler later after finishing with windows stuff(in a few days). I created an Issue already, "Reinforce RESP handler" 

![image](https://user-images.githubusercontent.com/8470490/198176642-9adf78f6-847e-4d41-b12b-e82148e4a637.png)

**2. Pipelining, streaming, and transactions**

These are buzzwords and have no relationship with each other. The sanest idea to improve buffer-related problems is to use a sync dispatch queue with if the pipeline is enabled, manage an integer ``write_index`` denoting "How far we have written in command buffer?", and keep incrementing it whenever we pop from the queue. i.e. if ``write_index`` is smaller than the size of the queue then keep doing ``pop_front`` until the evaluation of ``queue_size - write_index > 0`` holds true. If DB is in a transaction mode then the async dispatch queue could be a barrier to avoid out-of-order writes in the command buffer to maintain consistent transactions.

Reply:
Thanks @rohanverma94 and @asutosh97 keeping the existing logic as is then.

I was going for a strict RESP enforcement because I see supporting `+PING\r\n` is more of a legacy design choice that Redis is continuing to support but we can choose not to if it makes everything else better :)

Reply:
@rohanverma94 @arpitbbhayani I've again made some adjustments to the playground code to convey the bug I'm trying to address here: https://go.dev/play/p/gL-IIlQypgR

Please try running this code as it is once and after commenting the `if-block` highlighted in the code, hope this will clarify.

Reply:
@asutosh97 I'll address this issue. All problems boil down to the RESP handler. I'll demonstrate in an hour and probably to solve it we would need reactive buffer management. That's why it is better not to touch it.

Reply:
@asutosh97 Funny thing happened :) I spent the last 3.5 hours debugging this exact and trying to reproduce it.

What you proposed seems correct but when put into practice it is working just fine. 

**How I tested this?**
I took your example, set the buffer size to 7, and tested with your command pipelining input. I was shocked to see why it is not breaking. It should have. But for 1 pipelined `PING` it worked, for 2 also it worked for 3 also it did, till 50 it worked.

I even tested for other commands, different buffer lengths, different request sizes, and buffer sizes.

**why it is not failing**

Event Loop :) Since there is more data to be read from the socket the epoll_wait suggests that there are more bytes to be read. As I mentioned in the previous comment, the worst case is we would be doing multiple round trips for the evaluation of one command pipeline. (when buffer size is small) and that is expected.

This is the exact same behavior is observed.

Looks like we are safe here :) ~relieved~ 😌

Reply:
@arpitbbhayani yes ``var tmp []byte = make([]byte, 8)`` it is working completely fine, it all revolves around whether you read the complete command or not or you would just miss CRLF or even more than it. Check here https://rohanverma.godbolt.org/z/s31aP6P45

This however can break again in situations where you're limited by memory. To solve this we need reactive buffer management

Reply:
Do check my solution on this!

Reply:
I saw pending requests dropped. Here are the potential options

1. maintain a buffer per client and keep it handy until the client disconnects.
2. have a large buffer size ensuring we always read everything we need. and ignore the remaining buffer.

I am in favor of 1st approach though. But we will need to later take care of GC.

Reply:
@rohanverma94 and @asutosh97 let's address this in future commits. Is this PR good to go? or are there any major red flags?

Reply:
@arpitbbhayani Everything is good to go. For other issues as I've pointed out, we can create a bug report and work on that.

Reply:
@arpitbbhayani 
> why it is not failing
Event Loop :) Since there is more data to be read from the socket the epoll_wait suggests that there are more bytes to be read. As I mentioned in the previous comment, the worst case is we would be doing multiple round trips for the evaluation of one command pipeline. (when buffer size is small) and that is expected.

Wow! 🤯 
I had to check it for myself with the debugger turned on to see this in action, it feels like magic, this made me go back to the memories of learning duplex channels in communication, how on the same connection we are sending the response, and immediately after that reading the next set of inputs. The abstraction done by OS manages all that, it's really fascinating.

Okay, getting back to the review, this works for the cases I had thought it would break. But raises another concern:

```go
		if rp.buf.Len() > config.IOBufferLengthMAX {
			return nil, fmt.Errorf("input too long. max input can be %d bytes", config.IOBufferLengthMAX)
		}
```

This won't be effective anymore, a client can send data more than `config.IOBufferLengthMAX` and get away with it.
We can pick this up separately if you'd prefer that.


@rohanverma94 I'm unfamiliar about `reactive buffer management`, can you please explain it in simple terms?

Reply:
@asutosh97 let's pick this up separately. We anyway will need to write the correctness suite and ensure no memory leaks. I recommend picking this up with that.

Reply:
@asutosh97 As I've pointed out here https://github.com/DiceDB/dice/pull/52#discussion_r1008778089 that event loop magic may not be something that is overall good for the entire codebase. Reactive buffer management is a long shot, It would require a lot of refactoring and changes. The solution is to use "arena-based allocators", this would require converting the codebase to CGO. Read here - https://www.microsoft.com/en-us/research/uploads/prod/2019/06/mimalloc-tr-v1.pdf & https://microsoft.github.io/mimalloc/index.html

Reply:
@rohanverma94 agreed, anything magic seems good at first, but it makes it tricky when we are trying to debug something. Thanks for the documents, I'll go through them and reach out to you if I get have some doubts :)